### DISCUSSION TITLE
What is the best way to pass events to another Laravel project?

### DISCUSSION THREAD

**Discussion by PJKerrigan** [spatie/laravel-event-sourcing#291] [Q&A] [answered]
This is a great library. We're using it in production.

We've been asked to create a second Laravel project that will run on a different domain. It needs to generate reports based upon events recorded on the main platform. We'd then project data into the report tables so the users can pull out the exact data they need.

Do any of you know how best to pass events from one project to another?

Thanks.

**Comment by inmanturbo** [comment, marked as answer]
sounds like a good case for rabbitmq. Other options are to have your event project make an api call to the new project over http when events are fired. A similar but more scalable, "eventually consistent" way is you expose an "events" api in your current project then begin polling that api from your new project at intervals based on how quickly it will need the data, you could use artisan scheduler for this. What you are describing is essentially microservices.  Even still another, less advisable option is to poll the event stream directly over tcp using a shared database connection.

**Reply by brendt** [reply]
Agreed, you're looking for some kind of message bus that runs independently. Keep in mind that this package doesn't provide any infrastructure, you'd probably have to write a bridge that publishes local events on that global bus and listens for them as well. 

Luckily, in v7, you can create global event listeners by simply type hinting on `object`, or even an interface if you only want specific events to be published to your global bus. 

With regards to listening to global messages, you'll have to write the necessary code that either listens for global messages and publishes them as local events so that you can hook into them with this package, or does something more custom to your needs.

**Comment by renepardon** [comment]
I have a similar case because I use event sourcing within each microservice but want to publish domain events to a global event store (which is not yet in place) so all services can listen to that global/domain event.

I thought about using reactors to send specific events globally. Might this be a good idea or do you suggest something else? @inmanturbo

**Reply by inmanturbo** [reply]
if you are already running microservices with their own event stores and want to push to a global event store I would recommend Greg Young's [EventStoreDB](eventstore.com) if that is an option for you. There are cloud and open source versions. There is a client library for laravel here: https://github.com/Rxnet/eventstore-client and a wrapper for the client [here](https://github.com/gosuperscript/laravel-eventstore) which hooks into laravel's native event bus that might be worth looking into.

**Reply by renepardon** [reply]
Yep, there is also a Docker version of it. Question is just how to differenciate between "local" service events and the ones to push outside into global event store. Not all events might be usefull for global event store. Do you think a reactor might be the proper place to dispatch "global" events?

**Reply by inmanturbo** [reply]
I suppose the it would depend on whether you want to actually source these global events or just store them. If you just want to publish them globally from a reactor you could, but then you end up with two sources for the same event, and they could quickly become out of sync. Unless I'm misunderstanding you. Otherwise I suppose you would need an AR that takes in a connection flag. But an AR with more than one source is a scary thought.

**Reply by renepardon** [reply]
So I guess it might be a better idea to keep the event sourcing local within every service and just push domain/global events to rabbitmq through a reactor, so every other service can react on those global events.

**Reply by inmanturbo** [reply]
Yes rabbitmq would work for that. If you need a place to store global events, or aggregate the aggregates, so to speak, possibly for global reporting for example you could have a service read from the rabbitmq stream and project them out to a log or database. I wouldn't source them though. At least not from services that already have their own event store.
